presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,187 @@
+                    functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (functionResolution.isComparisonFunction(node.getFunctionHandle()) && node.getArguments().size() == 2) {
+                RowExpression left = node.getArguments().get(0);
+                RowExpression right = node.getArguments().get(1);
+                if (isCaseExpression(left)) {
+                    return processCaseExpression((SpecialFormExpression) left, right, node, false);
+                }
+                else if (isCaseExpression(right)) {
+                    return processCaseExpression((SpecialFormExpression) right, left, node, true);
+                }
+            }
+            return null;
+        }
+
+        private boolean isCaseExpression(RowExpression expression)
+        {
+            return expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm().equals(SWITCH);
+        }
+
+        private RowExpression processCaseExpression(SpecialFormExpression switchExpression,
+                                                    RowExpression valueExpression,
+                                                    CallExpression comparison,
+                                                    boolean isSwitchExpressionOnRight)
+        {
+            Optional<RowExpression> caseOperand = getCaseOperand(switchExpression);
+            List<RowExpression> whenClauses;
+            Optional<RowExpression> elseResult;
+
+            int argumentsSize = switchExpression.getArguments().size();
+            RowExpression last = switchExpression.getArguments().get(argumentsSize - 1);
+            if (last instanceof SpecialFormExpression && ((SpecialFormExpression) last).getForm().equals(WHEN)) {
+                whenClauses = switchExpression.getArguments().subList(1, argumentsSize);
+                elseResult = Optional.empty();
+            }
+            else {
+                whenClauses = switchExpression.getArguments().subList(1, argumentsSize - 1);
+                elseResult = Optional.of(last);
+            }
+
+            List<RowExpression> andExpressions = new ArrayList<>();
+            List<RowExpression> invertedOperands = new ArrayList<>();
+
+            for (RowExpression whenClause : whenClauses) {
+                RowExpression whenOperand = ((SpecialFormExpression) whenClause).getArguments().get(0);
+                RowExpression operand = caseOperand.map(x -> logicalRowExpressions.equals(x, whenOperand)).orElse(whenOperand);
+                RowExpression result = ((SpecialFormExpression) whenClause).getArguments().get(1);
+                RowExpression resultExpression = isSwitchExpressionOnRight ?
+                        replaceArguments(comparison, valueExpression, result) : replaceArguments(comparison, result, valueExpression);
```

Comment:
Add a helper method for this conditional so the logic is clear.

Reply:
Refactored the part.